File: src/netconsole/models/api/system_maintenance.py
from __future__ import annotations

from typing import Literal

from pydantic import Field, model_validator

from netconsole.models.api.common import ApiModel


class LogEntryDTO(ApiModel):
    time: str
    level: str
    display_level: str
    display_event: str
    display_detail: str
    raw_event: str
    raw_detail: str


class LogPageDTO(ApiModel):
    items: list[LogEntryDTO] = Field(default_factory=list)
    page: int
    page_size: int
    total: int
    total_pages: int


class RuntimeLogSummaryDTO(ApiModel):
    directory: str
    total_bytes: int
    protected_bytes: int
    unknown_bytes: int
    candidate_bytes: int
    protected_files: int
    unknown_files: int
    candidate_files: int
    max_total_bytes: int
    target_total_bytes: int
    retention_days: int


class CleanupItemDTO(ApiModel):
    item_id: str
    title: str
    description: str
    retention_policy: str
    status: str
    file_count: int
    total_bytes: int


class OpenSourceComponentDTO(ApiModel):
    name: str
    version: str
    license: str
    purpose: str
    homepage: str
    note: str


class MaintenanceTaskDTO(ApiModel):
    task_id: str
    status: str
    action: str
    progress: int = 0
    stage: str = ""
    message: str = ""
    error_message: str = ""
    artifact_id: str = ""
    artifact_kind: str = ""
    artifact_name: str = ""
    available: bool = False
    sha256: str = ""
    size_bytes: int = 0
    cleanup_items: list[CleanupItemDTO] = Field(default_factory=list)
    processed_files: int = 0
    deleted_files: int = 0
    failed_count: int = 0
    freed_bytes: int = 0
    deleted_log_records: int = 0
    scanned_log_records: int = 0
    malformed_log_records: int = 0
    rewritten_log_files: int = 0
    cutoff: str = ""
    components: list[OpenSourceComponentDTO] = Field(default_factory=list)


class CleanupStartRequest(ApiModel):
    mode: Literal["scan", "clean", "manual_history_cleanup"] = "scan"
    retention_days: int = Field(default=3, ge=1, le=365)
    selected_item_ids: list[Literal["runtime_logs", "runtime_cache", "temporary_files"]] = Field(
        default_factory=list,
        max_length=3,
    )
    confirmed: bool = False

    @model_validator(mode="after")
    def validate_cleanup_contract(self) -> "CleanupStartRequest":
        if len(self.selected_item_ids) != len(set(self.selected_item_ids)):
            raise ValueError("清理项目不能重复")
        if self.mode == "scan":
            if self.selected_item_ids or self.confirmed:
                raise ValueError("扫描请求只能包含保留天数")
            return self
        if self.mode == "manual_history_cleanup" and self.selected_item_ids != ["runtime_logs"]:
            raise ValueError("历史日志清理只允许选择 runtime_logs")
        if not self.selected_item_ids:
            raise ValueError("正式清理至少选择一个项目")
        if not self.confirmed:
            raise ValueError("正式清理必须明确确认")
        return self


class LogExportRequest(ApiModel):
    scope: Literal["current", "all"]
    keyword: str = Field(default="", max_length=200)
    level: Literal["", "INFO", "WARNING", "ERROR", "DEBUG", "CRITICAL"] = ""
    page: int = Field(default=1, ge=1, le=1_000_000)
    page_size: Literal[50, 100, 200, 500] = 200


class OpenSourceExportRequest(ApiModel):
    format: Literal["txt", "xlsx"]


class ChangelogDTO(ApiModel):
    title: str
    version: str
    content: str


class AboutLinkDTO(ApiModel):
    link_id: str
    label: str


class AboutDTO(ApiModel):
    title: str
    version: str
    author: str
    external_tool_notice: str
    repositories: list[AboutLinkDTO]


class DesktopActionDTO(ApiModel):
    success: bool
    code: str
    message: str = ""


class ExternalLinkDTO(ApiModel):
    url: str


__all__ = [
    "AboutDTO",
    "AboutLinkDTO",
    "ChangelogDTO",
    "CleanupItemDTO",
    "CleanupStartRequest",
    "DesktopActionDTO",
    "ExternalLinkDTO",
    "LogEntryDTO",
    "LogExportRequest",
    "LogPageDTO",
    "RuntimeLogSummaryDTO",
    "MaintenanceTaskDTO",
    "OpenSourceComponentDTO",
    "OpenSourceExportRequest",
]

File: apps/web/src/api/systemMaintenance.ts
import { apiRequest } from './client'

const root = '/api/system-maintenance'

export interface LogEntry {
  time: string
  level: string
  display_level: string
  display_event: string
  display_detail: string
  raw_event: string
  raw_detail: string
}

export interface LogPage {
  items: LogEntry[]
  page: number
  page_size: 50 | 100 | 200 | 500
  total: number
  total_pages: number
}

export type CleanupItemId = 'runtime_logs' | 'runtime_cache' | 'temporary_files'

export interface CleanupItem {
  item_id: CleanupItemId
  title: string
  description: string
  retention_policy: string
  status: string
  file_count: number
  total_bytes: number
}

export interface OpenSourceComponent {
  name: string
  version: string
  license: string
  purpose: string
  homepage: string
  note: string
}

export interface MaintenanceTask {
  task_id: string
  status: string
  action: string
  progress: number
  stage: string
  message: string
  error_message: string
  artifact_id: string
  artifact_kind: string
  artifact_name: string
  available: boolean
  sha256: string
  size_bytes: number
  cleanup_items: CleanupItem[]
  processed_files: number
  deleted_files: number
  failed_count: number
  freed_bytes: number
  components: OpenSourceComponent[]
}

export interface Changelog {
  title: string
  version: string
  content: string
}

export interface AboutInfo {
  title: string
  version: string
  author: string
  external_tool_notice: string
  repositories: Array<{ link_id: string; label: string }>
}

export function getLogs(query: { page: number; page_size: number; keyword: string; level: string }): Promise<LogPage> {
  return apiRequest<LogPage>(`${root}/logs?${new URLSearchParams(Object.entries(query).map(([key, value]) => [key, String(value)]))}`)
}

export function clearLogs(): Promise<{ success: boolean; code: string; message: string }> {
  return apiRequest(`${root}/logs`, { method: 'DELETE' })
}

export function startCleanup(payload: {
  mode: 'scan' | 'clean'
  retention_days: number
  selected_item_ids?: CleanupItemId[]
  confirmed?: boolean
}): Promise<MaintenanceTask> {
  return apiRequest(`${root}/cleanup/tasks`, { method: 'POST', body: JSON.stringify(payload) })
}

export function startOpenSourceScan(): Promise<MaintenanceTask> {
  return apiRequest(`${root}/open-source/tasks`, { method: 'POST', body: '{}' })
}

export function startLogExport(payload: {
  scope: 'current' | 'all'
  keyword: string
  level: string
  page: number
  page_size: number
}): Promise<MaintenanceTask> {
  return apiRequest(`${root}/exports/logs`, { method: 'POST', body: JSON.stringify(payload) })
}

export function startOpenSourceExport(format: 'txt' | 'xlsx'): Promise<MaintenanceTask> {
  return apiRequest(`${root}/exports/open-source`, { method: 'POST', body: JSON.stringify({ format }) })
}

export function getMaintenanceTask(taskId: string): Promise<MaintenanceTask> {
  return apiRequest(`${root}/tasks/${encodeURIComponent(taskId)}`)
}

export function recoverMaintenanceTasks(): Promise<MaintenanceTask[]> {
  return apiRequest(`${root}/tasks`)
}

export function cancelMaintenanceTask(taskId: string): Promise<MaintenanceTask> {
  return apiRequest(`${root}/tasks/${encodeURIComponent(taskId)}/cancel`, { method: 'POST', body: '{}' })
}

export function getChangelog(): Promise<Changelog> {
  return apiRequest(`${root}/changelog`)
}

export function getAbout(): Promise<AboutInfo> {
  return apiRequest(`${root}/about`)
}

export function requestAboutLink(linkId: string): Promise<{ url: string }> {
  return apiRequest(`${root}/links/about/${encodeURIComponent(linkId)}`, { method: 'POST', body: '{}' })
}

export function requestOpenSourceLink(taskId: string, componentIndex: number): Promise<{ url: string }> {
  return apiRequest(`${root}/links/open-source/${encodeURIComponent(taskId)}/${componentIndex}`, { method: 'POST', body: '{}' })
}

export function openMaintenanceDirectory(kind: 'logs' | 'cache'): Promise<{ success: boolean; code: string; message: string }> {
  return apiRequest(`${root}/desktop-actions/open-directory/${kind}`, { method: 'POST', body: '{}' })
}

export function maintenanceArtifactDownloadRequest(task: MaintenanceTask): { apiPath: string; suggestedName: string } {
  return {
    apiPath: `${root}/artifacts/${encodeURIComponent(task.artifact_kind)}/${encodeURIComponent(task.artifact_id)}`,
    suggestedName: task.artifact_name,
  }
}
